frontend/src/components/category-stats.tsx

```diff
@@ -0,0 +1,317 @@
+      codes.forEach(code => {
+        const stat = yearGroups[year][code];
+        if (stat?.mean_mark !== null && stat?.mean_mark !== undefined) {
+          yearData[code] = Number(stat.mean_mark.toFixed(1));
+          // Store standard deviation for tooltip
+          yearData[`${code}_std`] = stat?.std_deviation ? Number(stat.std_deviation.toFixed(1)) : null;
+          // Store organiser for tooltip
+          yearData[`${code}_organiser`] = stat.course_organiser;
+        }
+      });
+      return yearData;
+    });
+
+    // Generate reference lines for course organiser changes
+    const refLines: Array<{ x: string; label: string; color: string }> = [];
+    
+    codes.forEach(code => {
+      const courseStats = stats
+        .filter(s => s.course_code === code)
+        .sort((a, b) => a.academic_year.localeCompare(b.academic_year));
+
+      if (courseStats.length === 0) return;
+
+      // Add reference line for the initial organiser
+      const firstStat = courseStats[0];
+      if (firstStat?.course_organiser) {
+        const initialLabel = `${code}: ${firstStat.course_organiser}`;
+        const initialColor = courseOrganiserColors[firstStat.course_organiser] 
+          ? courseOrganiserColors[firstStat.course_organiser].replace('0.3', '0.8')
+          : 'gray.6';
+
+        refLines.push({
+          x: firstStat.academic_year,
+          label: initialLabel,
+          color: initialColor,
+        });
+      }
+
+      let previousOrganiser = firstStat?.course_organiser;
+
+      courseStats.forEach((stat, index) => {
+        if (index > 0 && stat.course_organiser !== previousOrganiser) {
+          // Course organiser changed in this year
+          const changeLabel = `${code}: ${stat.course_organiser}`;
+          const changeColor = courseOrganiserColors[stat.course_organiser || ''] 
+            ? courseOrganiserColors[stat.course_organiser || ''].replace('0.3', '0.8')
+            : 'gray.6';
+
+          // For the last year, position the label to the left to avoid overflow
+          const isLastYear = index === courseStats.length - 1;
+          
+          refLines.push({
+            x: stat.academic_year,
+            label: changeLabel,
+            color: changeColor,
+            labelPosition: isLastYear ? 'left' : 'right',
+          } as any); // Use 'as any' to bypass TypeScript check since this property works
+        }
+        previousOrganiser = stat.course_organiser;
+      });
+    });
+
+    return {
+      chartData: combinedData,
+      courseCodes: codes,
+      referenceLines: refLines,
+    };
+  }, [stats]);
+
+  // Colors for different course codes
+  const colors = [
+    "var(--mantine-primary-color-6)",
+    "var(--mantine-color-blue-6)",
+    "var(--mantine-color-green-6)",
+    "var(--mantine-color-yellow-6)",
+    "var(--mantine-color-red-6)",
+    "var(--mantine-color-violet-6)",
+    "var(--mantine-color-orange-6)",
+    "var(--mantine-color-teal-6)",
+  ];
+
+  if (loading) {
```

Comment:
I think it looks better if you change this condition to `loading && !stats`. When there is already a previous request by the same cacheKey, `stats` will be immediately populated and available, but `loading` will also be true, firing off a request for the most recent data.

So changing the condition will fix the jarring quick flash of loading skeletons when you switch between tabs that you already visited.